### PR TITLE
feat(workbox): workbox 3 + offlinePage

### PR DESCRIPTION
**Pull request by manniL**
Workbox has released version 3, so we could upgrade to v3 as well. But there are some obstacles.

In version 3, there is no `WorkboxSW` class anymore. `workbox` is exposed as a global variable instead (see the [Migration Guide](https://developers.google.com/web/tools/workbox/guides/migrations/migrate-from-v2#workbox-sw)). The problem is that `workbox` is always `undefined` when compiling the `sw.template.js` lodash template.

### CODE REVIEW

**Comment by manniL** [comment]
Unfortunately I don't know how to proceed further :( Hope if someone could give it a go and take over from here.

**Comment by XanderLuciano** [comment]
Just doing some preliminary research currently, but looks like vue-cli has a work in progress PWA with workbox 3. May be worth looking at the implementation there: https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-pwa

**Comment by monotykamary** [comment]
I think it might be related to the path of the `workbox-sw` asset.

[jsdelivr](https://www.jsdelivr.com/package/npm/workbox-sw) reflects the path for v2:

```
/build/importScripts/workbox-sw.prod.v2.1.3.min.js
```

and for v3:

```
/build/workbox-sw.min.js
```

File: packages/workbox/templates/sw.template.js [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+
+<% if(options.clientClaims) { %>
+  workbox.clientClaims()
+<% } %>
 
 <% options.runtimeCaching.forEach(r => {
   const strategy = JSON.stringify(r.strategyOptions || {})
   %>
-workboxSW.router.registerRoute(new RegExp('<%= r.urlPattern %>'), workboxSW.strategies.<%= r.handler %>(<%= strategy %>), '<%= r.method %>')
+workbox.routing.registerRoute(new RegExp('<%= r.urlPattern %>'), workboxSW.strategies.<%= r.handler %>(<%= strategy %>), '<%= r.method %>')
```
[the remark this answers is not in the document]
Comment (manniL) [reply packages/workbox/templates/sw.template.js]: Thanks, will fix!

**Comment by manniL** [comment]
Thanks to @XanderLuciano and @monotykamary for helping out :relaxed: 

Looks like the ball is rolling again :speak_no_evil: 

image

**Comment by manniL** [comment]
This looks somewhat ready. Will do some tests with a real-world project of mine and report back and soon as I have the results :relaxed:

**Comment by XanderLuciano** [comment]
Took me a bit to figure out how to install your changes locally but eventually figured it out and it seems to be working just fine on my project also!

image
image

**Comment by manniL** [comment]
@XanderLuciano  Happy to hear that!

The only thing I'm experiencing is that workbox is still used in dev mode/ on localhost, which results in an endless loop where workbox and the webpack HMR are fighting and reloading the page every few seconds. A hard refresh (CTRL + F5)  does stop it though.

**Comment by XanderLuciano** [comment]
@manniL I swear I had not run into this issue all night, but as soon as I saw your comment it happened. 

It seems a bit like an edge case for me though because in dev mode if I remove the service worker a new one isn't registered. So I was only able to reproduce by building and starting the app, then stopping the server and starting a dev instance, and then updating some code that triggers the HMR.

Like you said though, a quick ctrl+shift+r / ctrl + F5 fixed it. 

I'm gunna continue development using this version so if I run into additional issues (or a solution to the HMR reloading) I'll let you know.

**Comment by XanderLuciano** [comment]
Struggled to get offline support functional, but eventually figured it out.

```javascript
workbox: {
    runtimeCaching: [
        // Needed to cache the homepage and manifest file (not 100% sure on that though)
        {
            urlPattern: '/.*',
            handler: 'networkFirst',
            method: 'GET',
        },
        // Cache our CDN libraries
        {
            urlPattern: 'https://cdnjs.cloudflare.com/ajax/libs/.*',
            handler: 'cacheFirst',
            method: 'GET',
        },
        // Cache our fonts
        {
            urlPattern: 'https://fonts.(?:googleapis|gstatic).com/(.*)',
            strategyOptions: {
                cacheName: 'google-fonts',
            },
        },
    ],
},
```

Would it make sense to add that pattern to the template also? We already cache everything in `/_nuxt/`, but `/.*` would try to cache quite a bit more.

**Comment by manniL** [comment]
I've investigated the "update issue" and I've likely tracked it down. It is, as I assumed, a caching issue, so #63 should definitely be merged in before updating. After clearing my cache for localhost, everything worked well :)

**Comment by XanderLuciano** [comment]
@manniL Oh awesome! So merging #75 into my custom build of this PR should fix the issues with HMR? Have you tested offline usage with the server worker's cache disabled?

**Comment by XanderLuciano** [comment]
Well it took me 2 weeks to figure out how to create my own personal build of this package with all the PR's I wanted, but now I've got that all setup under my own personal scope (since it's taking so long to get any PR's to this repo). 

That said, I included #75 in that build and still get stuck in a reload loop when going from a built instance to a dev instance with hot reloading. So as you said, there are still some caveats to work out. 

For example, going from built version to dev, upon triggering an HMR the server console will output a loop of errors such as:

```
{
  statusCode: 404,
  path: '/_nuxt/55d3a1e8f0e84963b002.hot-update.json',
  message: 'This page could not be found'
}
```

And upon going from a dev instance to a built server instance, there appears to be an HMR request that continues to try and update, even though it's a production build:

image

So it seems to me that the overly broad URL pattern of `/.*` may be a little too broad and causing the service worker to cache these files. Potentially might be worth excluding files with `hot-update` in their name?

Maybe another solution to this would be to disable (or change)  precaching on development builds. That way when you you run `npm run dev` HMR works as expected and workbox doesn't interfere, but running `npm run build && npm run start` allows you test offline mode and your service worker properly.

I'm still learning a lot about workbox, but so far I've had no issues running this in a production environment. Add to homescreen banners appearing properly too.

**Comment by galvez** [comment]
@manniL @XanderLuciano anything else you'd like to get in before I merge this? I'll give it a round of testing tonight.

**Comment by manniL** [comment]
@galvez I’d wait until we fixed the plugin thing I mentioned above. Otherwise custom strategies are hard to accomplish. Testing is happily seen! 😋

**Comment by galvez** [comment]
@manniL I'm gonna dabble at fixing that problem, do you have any WIP updates to this branch?

**Comment by manniL** [comment]
@galvez Nope, have no WIP there yet :/

**Comment by galvez** [comment]
@manniL no worries -- picking this up on the `feature/workbox-v3` branch :)

**Comment by galvez** [comment]
@pi0 @manniL so just a quick update, I'm making some good progress with this -- just going through deploying an app to production with nuxt/pwa and am making notes as I go. I should get to it by the weekend. Looking forward to having you on the new Discord group @pi0! :)

**Comment by pi0** [comment]
@galvez Awesome news. I'm so happy having your help on the board. Waiting for PR 😍

**Comment by XanderLuciano** [comment]
Anyone running into any noteable bugs?

I've yet to run into any bugs (after caching was disabled) in production. 

I'd like to help resolve issues with any of the other workflows but I'm not aware of any issues. And if no one else knows of any issues it might be worthwhile updating and waiting for some members in the community to find bugs in the various workflow implementations.

**Comment by galvez** [comment]
@XanderLuciano almost there -- I'm adding script extension support (via `cachingExtensions` and `routingExtensions` and a way to specify an offline page.

**Comment by galvez** [comment]
@XanderLuciano https://discord.gg/VApZF5W :)

File: package.json [hunk 2]
```diff
@@ -7,7 +8,7 @@
     "test": "yarn lint && jest",
     "lerna": "lerna",
     "bootstrap": "lerna bootstrap",
-    "lint": "eslint --ext .js packages",
+    "lint": "$(npm bin)/eslint --ext .js packages",
```
[the remark this answers is not in the document]
Comment (pi0) [reply package.json]: I still don't get the point :) Both NPM and Yarn clients prepend.

Example:

package.json:

```
{
    "scripts" : {
        "test" : "echo $PATH",
        "test2" : "echo $(npm bin)"
    }
}
```

```bash
yarn test
/Users/pooya/tmp/a/node_modules/.bin:[other paths]

yarn test2
/Users/pooya/tmp/a/node_modules/.bin
```

Both are the same.

PS: this breaks windows support

File: packages/workbox/index.js [hunk 3]
```diff
@@ -166,7 +191,14 @@ function emitAssets (options) {
 function workboxInject (options) {
   const hook = () => {
     const opts = Object.assign({}, options)
+    const VALID_KEYS = [
```
Comment (pi0) [inline packages/workbox/index.js]: We can also use lodash `_.pick`

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (galvez) [reply package.json]: @pi0 you're right! Removed it :)

**Comment by manniL** [comment]
We should adopt the changes for `workbox-strategies` plugins as well. (see [here](https://developers.google.com/web/tools/workbox/guides/migrations/migrate-from-v2#workbox-strategies_formerly_know_as_workbox-runtime-caching))

**Comment by galvez** [comment]
> We should adopt the changes for workbox-strategies plugins as well. (see here)

As I have pointed out over Discord, we're not using any of the features that require updating.

**Pull request merged**